pkg/resource/predicates.go

```diff
@@ -72,3 +72,14 @@ func HasProvisioner(c client.Client, cr ClassReferencer, provisioner string) boo
 
 	return strings.EqualFold(cs.Provisioner, provisioner)
 }
+
+// NoClassReference accepts ResourceClaims that do not reference a specific ResourceClass
+func NoClassReference() PredicateFn {
```

Comment:
It occurred to me today during conversation with @bassam that we should probably further filter the claims watched by the default class controllers down to those that do not have a class reference _and_ do not have a managed resource reference.

There exists a use case in which a cluster operator might create, for example, a bunch of `RDSInstance` managed resources which could be later claimed by an application developer. Said developer would do this by submitting a `PostgreSQLInstance` claim with its `.spec.resourceRef` populated, referencing one of the existing `RDSInstance` managed resources. In this case there's no point setting a default resource class, because we're bypassing dynamic provisioning altogether.

I suggest we implement this functionality by:

1. Adding a `NoManagedResourceReference` variant of the existing `NoClassReference` function.
1. Updating the signature of `NewPredicates` to accept and apply multiple `PredicateFn`, e.g. `func NewPredicates(fn ...PredicateFn) predicate.Funcs`.
1. Updating the default class controllers to call `resource.NewPredicates(resource.NoClassReference(), resource.NoManagedResourceReference())`.

Reply:
This is a good point, and your suggestion for implementation follows what I would think to do as well. Thanks as always for your great feedback!

Reply:
My current implementation looks like this:
```
// NewPredicates returns a set of Funcs that are all satisfied by the supplied
// PredicateFn. The PredicateFn is run against the new object during updates.
func NewPredicates(fns ...PredicateFn) []predicate.Predicate {
	predicates := []predicate.Predicate{}
	for _, fn := range fns {
		predicate := predicate.Funcs{
			CreateFunc:  func(e event.CreateEvent) bool { return fn(e.Object) },
			DeleteFunc:  func(e event.DeleteEvent) bool { return fn(e.Object) },
			UpdateFunc:  func(e event.UpdateEvent) bool { return fn(e.ObjectNew) },
			GenericFunc: func(e event.GenericEvent) bool { return fn(e.Object) },
		}
		predicates = append(predicates, predicate)
	}

	return predicates
}
```

But this requires passing to controller watch like this:
```
resource.NewPredicates(resource.NoClassReference(), resource.NoManagedResourceReference())...
```
How do you feel about this pattern?

Reply:
Additional Note: this pattern requires adding `...` to the `NewPredicates()` calls in the claim controllers despite only specifying a single predicate.

Reply:
@hasheddan I had a slightly different implementation in mind, but did not realise that `c.Watch()` accepted variadic predicates. Given that it does, I'd instead just supply several `NewPredicates` like:

```go
	return errors.Wrapf(c.Watch(
		&source.Kind{Type: &databasev1alpha1.MySQLInstance{}},
		&handler.EnqueueRequestForObject{},
		resource.NewPredicates(resource.NoManagedResourceReference(),
		resource.NewPredicates(resource.NoClassReference()),
	), "cannot watch for %s", databasev1alpha1.MySQLInstanceGroupVersionKind)
```